### PR TITLE
Change AttributeMatrix/IGeometry Types Enumeration to std::vector from QVector

### PR DESCRIPTION
**Pull request by imikejackson**
Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

### CODE REVIEW

**Comment by imikejackson** [comment]
DREAM3DCI passed on all platforms
https://dev.azure.com/BlueQuartzSoftware/DREAM3D/_build/results?buildId=426&view=results

File: Source/SIMPLib/DataContainers/AttributeMatrix.cpp [hunk 1]
```diff
@@ -257,7 +257,9 @@ void AttributeMatrix::ReadAttributeMatrixStructure(hid_t containerId, DataContai
         if(req != nullptr)
         {
           AttributeMatrix::Types amTypes = req->getAMTypes();
-          if(amTypes.empty() || amTypes.contains(static_cast<AttributeMatrix::Type>(amTypeTmp)))
+          if(amTypes.empty() || std::find(amTypes.begin(), amTypes.end(), static_cast<AttributeMatrix::Type>(amTypeTmp)) != amTypes.end()
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/DataContainers/AttributeMatrix.cpp]:
```suggestion
          if(amTypes.empty() || std::find(amTypes.cbegin(), amTypes.cend(), static_cast<AttributeMatrix::Type>(amTypeTmp)) != amTypes.cend()
```

File: Source/SIMPLib/DataContainers/DataContainer.cpp [hunk 2]
```diff
@@ -132,19 +132,20 @@ void DataContainer::ReadDataContainerStructure(hid_t dcArrayGroupId, DataContain
...
+    uint32_t geometryType;
     herr_t err = QH5Lite::readScalarAttribute(containerGid, SIMPL::Geometry::Geometry, SIMPL::Geometry::GeometryType, geometryType);
     if(err >= 0)
     {
       dcProxy.setDCType(static_cast<unsigned int>(geometryType));
       if(req != nullptr)
       {
         IGeometry::Types geomTypes = req->getDCGeometryTypes();
-        if(geomTypes.empty() || geomTypes.contains(static_cast<IGeometry::Type>(geometryType)))
+        auto result = std::find(std::begin(geomTypes), std::end(geomTypes), static_cast<IGeometry::Type>(geometryType));
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/DataContainers/DataContainer.cpp]:
```suggestion
        auto result = std::find(geomTypes.cbegin(), geomTypes.cend(), static_cast<IGeometry::Type>(geometryType));
```

File: Source/SIMPLib/DataContainers/DataContainer.cpp [hunk 3]
```diff
@@ -132,19 +132,20 @@ void DataContainer::ReadDataContainerStructure(hid_t dcArrayGroupId, DataContain
...
+
+        if(geomTypes.empty() || result != std::end(geomTypes))
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/DataContainers/DataContainer.cpp]:
```suggestion
        if(geomTypes.empty() || result != geomTypes.cend())
```

File: Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp [hunk 4]
```diff
@@ -56,9 +56,10 @@ class FilterParameterWidgetUtils
       }
       dcCombo->addItem(dcProxy.getName());
 
-      if(!geomTypes.isEmpty())
+      if(!geomTypes.empty())
       {
-        if(!geomTypes.contains(geomType))
+        if(std::find(std::begin(geomTypes), std::end(geomTypes), static_cast<IGeometry::Type>(geomType)) == std::end(geomTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp]:
```suggestion
        if(std::find(geomTypes.cbegin(), geomTypes.cend(), static_cast<IGeometry::Type>(geomType)) == geomTypes.cend())
```

File: Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp [hunk 5]
```diff
@@ -56,9 +56,10 @@ class FilterParameterWidgetUtils
...
+        // if(!geomTypes.contains(geomType))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp]:
```suggestion
```

File: Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp [hunk 6]
```diff
@@ -124,7 +125,9 @@ class FilterParameterWidgetUtils
           AttributeMatrix::Pointer am = dca->getAttributeMatrix(DataArrayPath(dc.getName(), amName, ""));
           amCombo->addItem(amName);
 
-          if(nullptr != am.get() && !defVec.isEmpty() && !defVec.contains(am->getType()))
+          auto result = std::find(std::begin(defVec), std::end(defVec), static_cast<AttributeMatrix::Type>(am->getType()));
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp]:
```suggestion
          auto result = std::find(defVec.cbegin(), defVec.cend(), static_cast<AttributeMatrix::Type>(am->getType()));
```

File: Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp [hunk 7]
```diff
@@ -124,7 +125,9 @@ class FilterParameterWidgetUtils
...
+
+          if(nullptr != am.get() && !defVec.empty() && result == std::end(defVec))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/FilterParameterWidgets/FilterParameterWidgetUtils.hpp]:
```suggestion
          if(nullptr != am.get() && !defVec.empty() && result == defVec.cend())
```

File: Source/SVWidgetsLib/FilterParameterWidgets/LinkedDataContainerSelectionWidget.cpp [hunk 8]
```diff
@@ -215,7 +215,10 @@ void LinkedDataContainerSelectionWidget::createSelectionMenu()
     m_MenuMapper->setMapping(action, path);
     menu->addAction(action);
 
-    if(!geomTypes.isEmpty() && !geomTypes.contains(geomType) && !geomTypes.contains(IGeometry::Type::Any))
+    auto result = std::find(std::begin(geomTypes), std::end(geomTypes), static_cast<IGeometry::Type>(geomType));
+    auto result1 = std::find(std::begin(geomTypes), std::end(geomTypes), IGeometry::Type::Any);
+
+    if(!geomTypes.empty() && result == std::end(geomTypes) && result1 == std::end(geomTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/FilterParameterWidgets/LinkedDataContainerSelectionWidget.cpp]:
```suggestion
    auto result = std::find(geomTypes.cbegin(), geomTypes.cend(), static_cast<IGeometry::Type>(geomType));
    auto result1 = std::find(geomTypes.cbegin(), geomTypes.cend(), IGeometry::Type::Any);

    if(!geomTypes.empty() && result == geomTypes.cend() && result1 == geomTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 9]
```diff
@@ -199,14 +199,19 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
   if(!reqs.dcGeometryTypes.empty())
   {
     // Unknown Geometry gets a pass
-    if(!reqs.dcGeometryTypes.contains(IGeometry::Type::Unknown))
+    auto result1 = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), IGeometry::Type::Unknown);
+    if(result1 == std::end(reqs.dcGeometryTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
    auto result1 = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), IGeometry::Type::Unknown);
    if(result1 == reqs.dcGeometryTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 10]
```diff
@@ -199,14 +199,19 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
...
       if(nullptr == geom)
       {
         return false;
       }
-      if(!reqs.dcGeometryTypes.contains(geom->getGeometryType()) && !reqs.dcGeometryTypes.contains(IGeometry::Type::Any))
+
+      auto result = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), geom->getGeometryType());
+      auto result2 = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), IGeometry::Type::Any);
+
+      if(result == std::end(reqs.dcGeometryTypes) && result2 == std::end(reqs.dcGeometryTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
      auto result = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), geom->getGeometryType());
      auto result2 = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), IGeometry::Type::Any);

      if(result == reqs.dcGeometryTypes.cend() && result2 == reqs.dcGeometryTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 11]
```diff
@@ -245,14 +250,18 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
   if(!reqs.dcGeometryTypes.empty())
   {
     // Unknown Geometry gets a pass
-    if(!reqs.dcGeometryTypes.contains(IGeometry::Type::Unknown))
+    auto result1 = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), IGeometry::Type::Unknown);
+    if(result1 == std::end(reqs.dcGeometryTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
    auto result1 = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), IGeometry::Type::Unknown);
    if(result1 == reqs.dcGeometryTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 12]
```diff
@@ -245,14 +250,18 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
...
       IGeometry::Pointer geom = dc->getGeometry();
       if(nullptr == geom)
       {
         return false;
       }
-      if(!reqs.dcGeometryTypes.contains(geom->getGeometryType()) && !reqs.dcGeometryTypes.contains(IGeometry::Type::Any))
+      auto result = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), geom->getGeometryType());
+      auto result2 = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), IGeometry::Type::Any);
+
+      if(result == std::end(reqs.dcGeometryTypes) && result2 == std::end(reqs.dcGeometryTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
      auto result = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), geom->getGeometryType());
      auto result2 = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), IGeometry::Type::Any);

      if(result == reqs.dcGeometryTypes.cend() && result2 == reqs.dcGeometryTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 13]
```diff
@@ -266,9 +275,13 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
     return false;
   }
 
-  if(!reqs.amTypes.contains(AttributeMatrix::Type::Unknown))
+  auto result = std::find(std::begin(reqs.amTypes), std::end(reqs.amTypes), AttributeMatrix::Type::Unknown);
+
+  if(result == std::end(reqs.amTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
  auto result = std::find(reqs.amTypes.cbegin(), reqs.amTypes.cend(), AttributeMatrix::Type::Unknown);

  if(result == reqs.amTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 14]
```diff
@@ -266,9 +275,13 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
...
   {
-    if(!reqs.amTypes.empty() && !(reqs.amTypes.contains(am->getType()) || reqs.amTypes.contains(AttributeMatrix::Type::Any)))
+    auto result1 = std::find(std::begin(reqs.amTypes), std::end(reqs.amTypes), am->getType());
+    auto result2 = std::find(std::begin(reqs.amTypes), std::end(reqs.amTypes), AttributeMatrix::Type::Any);
+    if(!reqs.amTypes.empty() && result1 == std::end(reqs.amTypes) && result2 == std::end(reqs.amTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
    auto result1 = std::find(reqs.amTypes.cbegin(), reqs.amTypes.cend(), am->getType());
    auto result2 = std::find(reqs.amTypes.cbegin(), reqs.amTypes.cend(), AttributeMatrix::Type::Any);
    if(!reqs.amTypes.empty() && result1 == reqs.amTypes.cend() && result2 == reqs.amTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 15]
```diff
@@ -319,14 +332,19 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
   if(!reqs.dcGeometryTypes.empty())
   {
     // Unknown Geometry gets a pass
-    if(!reqs.dcGeometryTypes.contains(IGeometry::Type::Unknown))
+    auto result = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), IGeometry::Type::Unknown);
+
+    if(result == std::end(reqs.dcGeometryTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
    auto result = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), IGeometry::Type::Unknown);

    if(result == reqs.dcGeometryTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 16]
```diff
@@ -319,14 +332,19 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
...
       IGeometry::Pointer geom = dc->getGeometry();
       if(nullptr == geom)
       {
         return false;
       }
-      if(!reqs.dcGeometryTypes.contains(geom->getGeometryType()) && !reqs.dcGeometryTypes.contains(IGeometry::Type::Any))
+
+      auto result1 = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), geom->getGeometryType());
+      auto result2 = std::find(std::begin(reqs.dcGeometryTypes), std::end(reqs.dcGeometryTypes), IGeometry::Type::Any);
+      if(result1 == std::end(reqs.dcGeometryTypes) && result2 == std::end(reqs.dcGeometryTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
      auto result1 = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), geom->getGeometryType());
      auto result2 = std::find(reqs.dcGeometryTypes.cbegin(), reqs.dcGeometryTypes.cend(), IGeometry::Type::Any);
      if(result1 == reqs.dcGeometryTypes.cend() && result2 == reqs.dcGeometryTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 17]
```diff
@@ -340,9 +358,13 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
     return false;
   }
 
-  if(!reqs.amTypes.contains(AttributeMatrix::Type::Unknown))
+  auto result = std::find(std::begin(reqs.amTypes), std::end(reqs.amTypes), AttributeMatrix::Type::Unknown);
+
+  if(result == std::end(reqs.amTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
  auto result = std::find(reqs.amTypes.cbegin(), reqs.amTypes.cend(), AttributeMatrix::Type::Unknown);

  if(result == reqs.amTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 18]
```diff
@@ -340,9 +358,13 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
...
   {
-    if(!reqs.amTypes.empty() && !reqs.amTypes.contains(am->getType()) && !reqs.amTypes.contains(AttributeMatrix::Type::Any))
+    auto result1 = std::find(std::begin(reqs.amTypes), std::end(reqs.amTypes), am->getType());
+    auto result2 = std::find(std::begin(reqs.amTypes), std::end(reqs.amTypes), AttributeMatrix::Type::Any);
+    if(!reqs.amTypes.empty() && result1 == std::end(reqs.amTypes) && result2 == std::end(reqs.amTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
    auto result1 = std::find(reqs.amTypes.cbegin(), reqs.amTypes.cend(), am->getType());
    auto result2 = std::find(reqs.amTypes.cbegin(), reqs.amTypes.cend(), AttributeMatrix::Type::Any);
    if(!reqs.amTypes.empty() && result1 == reqs.amTypes.cend() && result2 == reqs.amTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp [hunk 19]
```diff
@@ -355,7 +377,8 @@ bool DataArrayPathSelectionWidget::CheckPathRequirements(AbstractFilter* filter,
     return false;
   }
 
-  if(!reqs.daTypes.empty() && !reqs.daTypes.contains(da->getTypeAsString()))
+  auto result3 = std::find(std::begin(reqs.daTypes), std::end(reqs.daTypes), da->getTypeAsString());
+  if(result3 == std::end(reqs.daTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/DataArrayPathSelectionWidget.cpp]:
```suggestion
  auto result3 = std::find(reqs.daTypes.cbegin(), reqs.daTypes.cend(), da->getTypeAsString());
  if(result3 == reqs.daTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/ImportASCIIDataWizard/DataFormatPage.cpp [hunk 20]
```diff
@@ -473,7 +473,10 @@ void DataFormatPage::createAMSelectionMenu()
     QMenu* dcMenu = btnMenu->addMenu(dc->getName()); // BtnMenu owns the new QMenu
     dcMenu->setDisabled(false);
 
-    if(!geomTypes.isEmpty() && !geomTypes.contains(geomType) && !geomTypes.contains(IGeometry::Type::Any))
+    auto result = std::find(std::begin(geomTypes), std::end(geomTypes), static_cast<IGeometry::Type>(geomType));
+    auto result1 = std::find(std::begin(geomTypes), std::end(geomTypes), IGeometry::Type::Any);
+
+    if(!geomTypes.empty() && result == std::end(geomTypes) && result1 == std::end(geomTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/ImportASCIIDataWizard/DataFormatPage.cpp]:
```suggestion
    auto result = std::find(geomTypes.cbegin(), geomTypes.cend(), static_cast<IGeometry::Type>(geomType));
    auto result1 = std::find(geomTypes.cbegin(), geomTypes.cend(), IGeometry::Type::Any);

    if(!geomTypes.empty() && result == geomTypes.cend() && result1 == geomTypes.cend())
```

File: Source/SVWidgetsLib/Widgets/ImportASCIIDataWizard/DataFormatPage.cpp [hunk 21]
```diff
@@ -493,7 +496,8 @@ void DataFormatPage::createAMSelectionMenu()
       m_AMMenuMapper->setMapping(amAction, path);
 
       bool amIsNotNull = nullptr != am.get();
-      bool amValidType = !amTypes.isEmpty() && !amTypes.contains(am->getType());
+      auto result2 = std::find(std::begin(amTypes), std::end(amTypes), am->getType());
+      bool amValidType = !amTypes.empty() && result2 == std::end(amTypes);
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/ImportASCIIDataWizard/DataFormatPage.cpp]:
```suggestion
      auto result2 = std::find(amTypes.cbegin(), amTypes.cend(), am->getType());
      bool amValidType = !amTypes.empty() && result2 == amTypes.cend();
```

File: Source/SVWidgetsLib/Widgets/ImportASCIIDataWizard/DataFormatPage.cpp [hunk 22]
```diff
@@ -560,7 +564,10 @@ void DataFormatPage::createDCSelectionMenu()
     connect(dcAction, SIGNAL(triggered(bool)), m_DCMenuMapper, SLOT(map()));
     m_DCMenuMapper->setMapping(dcAction, path);
 
-    if(!geomTypes.isEmpty() && !geomTypes.contains(geomType) && !geomTypes.contains(IGeometry::Type::Any))
+    auto result = std::find(std::begin(geomTypes), std::end(geomTypes), static_cast<IGeometry::Type>(geomType));
+    auto result1 = std::find(std::begin(geomTypes), std::end(geomTypes), IGeometry::Type::Any);
+
+    if(!geomTypes.empty() && result == std::end(geomTypes) && result1 == std::end(geomTypes))
```
Comment (JDuffeyBQ) [inline Source/SVWidgetsLib/Widgets/ImportASCIIDataWizard/DataFormatPage.cpp]:
```suggestion
    auto result = std::find(geomTypes.cbegin(), geomTypes.cend(), static_cast<IGeometry::Type>(geomType));
    auto result1 = std::find(geomTypes.cbegin(), geomTypes.cend(), IGeometry::Type::Any);

    if(!geomTypes.empty() && result == geomTypes.cend() && result1 == geomTypes.cend())
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]
It may be a good idea to implement our own contains function since we use it a lot.

```
namespace SIMPL
{
template <class Iterator, class T>
constexpr bool contains(Iterator first, Iterator last, const T& value)
{
  // Can't use std::find in the implemention for constexpr until C++20

  for(; first != last; ++first)
  {
    if(*first == value)
    {
      return true;
    }
  }

  return false;
}
} // namespace SIMPL
```

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**